### PR TITLE
[fsdp] fix: sync merged LoRA weights before context exit

### LINKED ISSUE #6782: [fsdp, rollout] Qwen3.5-27B LoRA GRPO: vLLM never emits EOS with rank=64/alpha=128 (works with rank=32/alpha=64), all responses truncated
### System Info

----------Python Info----------
Version      : 3.12.3
Compiler     : GCC 13.3.0
Build        : ('main', )
Arch         : ('64bit', '')
------------Pip Info-----------
Version      : 26.1.1
Directory    : /usr/local/lib/python3.12/dist-packages/pip
vllm         : 0.20.2
sglang       : not found.
ray          : 2.55.1
torch        : 2.11.0+cu130
----------verl Info-----------
Version      : 0.9.0.dev
Directory    : verl
Commit Hash  : 7213069ab418ec4be3d209c57e9959d87a462174
----------Environment----------
CUDA Runtime : 13.0
CUDA Compiler : Cuda compilation tools, release 13.0, V13.0.88
----------System Info----------
CPU Memory      : 2014.74 GB
GPU Count       : 8
GPU 1   Type    : NVIDIA H20
GPU 1   Memory  : 95.58 GB
GPU 2   Type    : NVIDIA H20
GPU 2   Memory  : 95.58 GB
GPU 3   Type    : NVIDIA H20
GPU 3   Memory  : 95.58 GB
GPU 4   Type    : NVIDIA H20
GPU 4   Memory  : 95.58 GB
GPU 5   Type    : NVIDIA H20
GPU 5   Memory  : 95.58 GB
GPU 6   Type    : NVIDIA H20
GPU 6   Memory  : 95.58 GB
GPU 7   Type    : NVIDIA H20
GPU 7   Memory  : 95.58 GB
GPU 8   Type    : NVIDIA H20
GPU 8   Memory  : 95.58 GB

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

docker from : verlai/verl:vllm020.dev2

set -xeuo pipefail

export CUDA_DEVICE_MAX_CONNECTIONS=1
export VLLM_USE_V1=1
export VLLM_ALLREDUCE_USE_SYMM_MEM=0

# ---- 保存地址 ----
EXPERIMENT_NAME=qwen3_5_27b_lora_fsdp
PROJECT_NAME=verl_grpo_lora_gsm8k
OUTPUT_DIR=workspace/train_ckpt/${EXPERIMENT_NAME}   # 模型/检查点保存地址
LOCAL_RUN_DIR=runs/${EXPERIMENT_NAME}                # 运行日志地址
export SWANLAB_LOG_DIR=/xrk_core/runs/swanwab
export SWANLAB_MODE=local
start_time=$(date +%Y%m%d)_$(date +%H%M%S)
mkdir -p "${LOCAL_RUN_DIR}/logs" "${OUTPUT_DIR}" "${SWANLAB_LOG_DIR}"

python3 -m verl.trainer.main_ppo \
    algorithm.adv_estimator=grpo \
    algorithm.use_kl_in_reward=False \
    data.train_files=workspace/train_data/math_data/gsm8k/train.parquet \
    data.val_files=workspace/train_data/math_data/gsm8k/test.parquet \
    data.train_batch_size=64 \
    data.max_prompt_length=1024 \
    data.max_response_length=1024 \
    data.filter_overlong_prompts=True \
    data.truncation=error \
    actor_rollout_ref.model.path=workspace/train_model/Qwen3.5-27B \
    actor_rollout_ref.model.trust_remote_code=True \
    actor_rollout_ref.model.use_remove_padding=False \
    actor_rollout_ref.model.enable_gradient_checkpointing=True \
    actor_rollout_ref.model.lora_rank=64 \
    actor_rollout_ref.model.lora_alpha=128 \
    actor_rollout_ref.model.target_modules=all-linear \
    actor_rollout_ref.model.exclude_modules='.*visual.*' \
    actor_rollout_ref.model.lora.merge=False \
    actor_rollout_ref.actor.strategy=fsdp2 \
    actor_rollout_ref.actor.optim.lr=1e-5 \
    actor_rollout_ref.actor.ppo_mini_batch_size=32 \
    actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=1 \
    actor_rollout_ref.actor.use_dynamic_bsz=False \
    actor_rollout_ref.actor.use_kl_loss=True \
    actor_rollout_ref.actor.kl_loss_coef=0.001 \
    actor_rollout_ref.actor.kl_loss_type=low_var_kl \
    actor_rollout_ref.actor.entropy_coeff=0 \
    actor_rollout_ref.actor.fsdp_config.param_offload=True \
    actor_rollout_ref.actor.fsdp_config.optimizer_offload=True \
    actor_rollout_ref.rollout.name=vllm \
    actor_rollout_ref.rollout.tensor_model_parallel_size=4 \
    actor_rollout_ref.rollout.gpu_memory_utilization=0.5 \
    actor_rollout_ref.rollout.enforce_eager=True \
    actor_rollout_ref.rollout.free_cache_engine=True \
    actor_rollout_ref.rollout.n=5 \
    actor_rollout_ref.rollout.dtype=bfloat16 \
    actor_rollout_ref.rollout.load_format=safetensors \
    actor_rollout_ref.rollout.layered_summon=True \
    actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=1 \
    actor_rollout_ref.rollout.log_prob_use_dynamic_bsz=False \
    actor_rollout_ref.ref.strategy=fsdp2 \
    actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=1 \
    actor_rollout_ref.ref.log_prob_use_dynamic_bsz=False \
    actor_rollout_ref.ref.fsdp_config.param_offload=True \
    trainer.critic_warmup=0 \
    trainer.balance_batch=True \
    trainer.logger='["console","swanlab"]' \
    trainer.project_name=${PROJECT_NAME} \
    trainer.experiment_name=${EXPERIMENT_NAME} \
    trainer.n_gpus_per_node=8 \
    trainer.nnodes=1 \
    trainer.save_freq=10 \
    trainer.test_freq=10 \
    trainer.val_before_train=False \
    trainer.total_epochs=15 \
    trainer.resume_mode=auto \
    trainer.max_actor_ckpt_to_keep=5 \
    trainer.default_local_dir=${OUTPUT_DIR} \
    "$@" 2>&1 | tee "${LOCAL_RUN_DIR}/logs/${EXPERIMENT_NAME}-${start_time}.log"

## Background — what changed on `main`

Qwen3.5 support landed on `main` through a series of PRs:

- **#5682** `[fsdp, model] feat: add qwen3.5 fsdp grpo training support` — first enabled FSDP GRPO for `Qwen3_5ForConditionalGeneration`.
- **#6482** `[fsdp, model] feat: support qwen3_5 ulysses sp`.
- **#5599** `[megatron] Qwen3.5 LoRA & MTP support (Megatron-Bridge)` — current `HEAD` (`7213069a`).
- **#6688** (fixes **#6454**) `[rollout] fix: clone LoRA weights out of the reused IPC buffer before add_lora` — fixed `cudaErrorIllegalAddress` for unmerged LoRA (`lora.merge=False`) + vLLM + `free_cache_engine=True`.

I'm on `HEAD = 7213069a`, running **Qwen3.5-27B LoRA GRPO on GSM8K**, FSDP2 actor/ref + vLLM rollout (TP=4), unmerged dynamic LoRA (`lora.merge=False`, `load_format=safetensors`, `layered_summon=True`). Running inside the **official verl Docker image**.

## What happens

The **only** difference between the two runs is the LoRA rank/alpha:

| Run | `lora_rank` | `lora_alpha` | Result |
|-----|-------------|--------------|--------|
| A | 32 | 64 | ✅ Works. vLLM emits EOS normally, responses terminate, rewards/score behave as expected. |
| B | 64 | 128 | ❌ vLLM rollouts **never emit EOS**; every response runs to `max_response_length` and is truncated (`response_length/clip_ratio ≈ 1.0`), so scores collapse. |

Everything else (model, data, batch sizes, sampling params, `max_prompt_length=1024`, `max_response_length=1024`, TP, offload, merge/layered_summon settings) is identical between A and B.

## Expected behavior

Increasing the LoRA rank/alpha from `rank=32, alpha=64` to `rank=64, alpha=128` should only change the capacity/scale of the LoRA adapters; it should **not** alter the rollout's ability to terminate generation. With `rank=64, alpha=128` I expect vLLM to keep producing well-formed responses that emit the EOS token and stop before `max_response_length`, with `response_length/clip_ratio` staying low — i.e. the same qualitative rollout behavior as the working `rank=32, alpha=64` run, just with a higher-capacity adapter. Training should proceed with valid, non-degenerate reward signal rather than every sample being truncated.

### PR DESCRIPTION
**Pull request by rongkunxue**
This PR fixes a stale-weight bug in the FSDP LoRA rollout weight sync path when `actor_rollout_ref.model.lora.merge=True`.

Before this change, `FSDPEngine.get_per_tensor_param()` built a `state_dict()` inside `merged_lora_context()`, but returned a lazy iterator that was consumed only after the context had already exited. Since `state_dict()` tensors alias the live module parameter storage, and `merged_lora_context()` restores the un-merged base weights on exit, the rollout engine could silently receive base weights without the LoRA delta.

As a result, vLLM rollout weights stopped tracking the latest actor LoRA updates after the first step. In affected runs, `training/rollout_probs_diff_mean` kept increasing step by step, while `rollout_actor_probs_pearson_corr` gradually dropped and rewards did not improve.

This PR keeps the merged LoRA context alive until every tensor is materialized and yielded to the rollout weight sync consumer.

Relevant issue: https://github.com/verl-project/verl/issues/6782

This does not duplicate an existing PR because issue #6782 currently has no linked PR on GitHub, and the following duplicate-work searches did not show an existing open PR for the same fix:

- https://github.com/verl-project/verl/pulls?q=is%3Apr+is%3Aopen+6782
- https://github.com/verl-project/verl/pulls?q=is%3Apr+is%3Aopen+merged_lora_context
- https://github.com/verl-project/verl/pulls?q=is%3Apr+is%3Aopen+lora+merge+rollout+weight+sync

### Test

This change has been tested and verified to resolve [#6782](https://github.com/verl-project/verl/issues/6782).

- `PATH="$PWD/.venv/bin:$PATH" .venv/bin/pre-commit run --all-files --show-diff-on-failure --color=never`
- `pytest -q tests/workers/test_engine_workers_lora_sync.py` (16 passed)

Full reproduction logs, validation results, and training behavior will be synchronized to the corresponding issue.

### API and Usage Example

No API or config changes.

The existing `actor_rollout_ref.model.lora.merge=True` path now sends the merged base + LoRA weights correctly during rollout weight sync.

### Design & Code Changes

The root cause is that the previous merge path created a `state_dict()` inside `merged_lora_context()`, but materialized the tensors only after leaving the context. Since the context restores the original base weights on exit, the lazy consumer could receive restored base weights instead of merged LoRA weights.

This PR changes the merged LoRA path to stream weights through a dedicated generator:

- Keep `merged_lora_context(self.module, backup_adapters=True)` open inside the generator.
- Normalize and convert weight keys while the merged weights are active.
- Materialize `DTensor` values with `full_tensor()` before yielding them.
- Clone plain tensors before yielding to avoid aliasing with restored module storage.
- Move FSDP parameter offload into a `finally` block so offload happens after the generator is fully consumed.

The unmerged LoRA path (`merge=False`) and non-LoRA path are unchanged.

AI assistance was used to help draft this PR description. The human submitter reviewed every changed line, validated the behavior, and is responsible for the final PR.

### CODE REVIEW

File: verl/workers/engine/fsdp/transformer_impl.py [hunk 1]
```diff
@@ -823,14 +820,7 @@ def get_per_tensor_param(self, layered_summon=False, base_sync_done=False, **kwa
...
         # leaves the module half-moved and crashes state_dict() below (#5995). The
         # per-DTensor .to(device).full_tensor() below still produces GPU tensors.
-        #
-        # FSDP2 state_dict() only collects DTensor refs and the generator below already
-        # stages each shard lazily via .to(device).full_tensor(), so the whole-shard
-        # round trip is only needed for FSDP1 (state_dict unshards on-device) and LoRA
-        # (adapter merge does real weight math on the module).
-        _is_peft = hasattr(getattr(self.module, "_fsdp_wrapped_module", self.module), "peft_config")
-        _skip_staging = fsdp_version(self.module) == 2 and not _is_peft
-        if not self._uses_fsdp2_cpu_offload_policy and not _skip_staging:
```
Comment (wuxibin89) [inline verl/workers/engine/fsdp/transformer_impl.py:833]: Why remove this? https://github.com/verl-project/verl/pull/7005

File: verl/workers/engine/fsdp/transformer_impl.py [hunk 2]
```diff
@@ -668,9 +668,6 @@ def forward_backward_batch(self, data: TensorDict, loss_function: Callable, forw
                         scaler.scale(loss).backward()
                     else:
                         loss.backward()
-                    # Training discards model_output (train_batch pops it); keeping it accumulates
-                    # full-length nested tensors across the mini-batch (∝ ppo_mini_batch * rollout_n) → OOM.
-                    meta_info.pop("model_output", None)
```
Comment (wuxibin89) [inline verl/workers/engine/fsdp/transformer_impl.py:673]: For training, the `model_output` is popped in later batch processing, so there's no need to keep it here.
https://github.com/verl-project/verl/blob/main/verl/workers/engine_workers.py#L364-L365

In which case do you need to consume this `model_output`?

Comment (rongkunxue) [reply verl/workers/engine/fsdp/transformer_impl.py:673]: Thanks, both points are valid. Fixed in ee5d95b7.

File: verl/workers/engine/fsdp/transformer_impl.py [hunk 3]
```diff
@@ -850,33 +851,36 @@ def get_per_tensor_param(self, layered_summon=False, base_sync_done=False, **kwa
...
                     params = {replace_lora_wrapper(k, peft_config): v for k, v in params.items()}
             else:  # merge lora
-                with merged_lora_context(self.module, backup_adapters=True):
-                    params = self.module.state_dict()
-                    params = normalize_peft_param_name(params)
+                # state_dict() aliases the live parameter storage and merged_lora_context
+                # restores the un-merged base weights on exit, so tensors must be
+                # materialized while the context is still open (inside the generator).
+                # Materializing after exit silently sends base weights without adapters.
+                per_tensor_param = self._merged_lora_per_tensor_param()
```
Comment (wuxibin89) [inline verl/workers/engine/fsdp/transformer_impl.py]: Better to return here, no need to change below.

Comment (rongkunxue) [reply verl/workers/engine/fsdp/transformer_impl.py]: Do you mean returning self._merged_lora_per_tensor_param(), None directly here? My concern is that an early return would bypass the shared QAT handling below. The current guard skips the original params processing for merged LoRA while still allowing per_tensor_param to go through QAT. If merged LoRA with QAT is not supported, I can use the early return and leave L862–L882 unchanged.

Comment (wuxibin89) [reply verl/workers/engine/fsdp/transformer_impl.py]: We're not going to support LoRA+QAT

Comment (rongkunxue) [reply verl/workers/engine/fsdp/transformer_impl.py]: Thanks, Updated in ff6724da.

**Review by wuxibin89 — APPROVED** [review]
